tolstoy/src/syncer.rs

```diff
@@ -155,20 +155,70 @@ impl Syncer {
             // On succes:
             // - persist local mappings from the receiver
             // - update our local "remote head".
-            TxMapper::set_bulk(&mut db_tx, &uploader.tx_temp_uuids)?;
-            SyncMetadataClient::set_remote_head(&db_tx, &last_tx_uploaded)?;
+            TxMapper::set_bulk(db_tx, &uploader.tx_temp_uuids)?;
+            SyncMetadataClient::set_remote_head(db_tx, &last_tx_uploaded)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn flow(sqlite: &mut rusqlite::Connection, server_uri: &String, user_uuid: &Uuid) -> Result<()> {
+        d(&format!("sync flowing"));
 
-            // Commit everything: tx->uuid mappings and the new HEAD. We're synced!
-            db_tx.commit()?;
+        ensure_current_version(sqlite)?;
+        
+        // TODO configure this sync with some auth data
+        let remote_client = RemoteClient::new(server_uri.clone(), user_uuid.clone());
+        let mut db_tx = sqlite.transaction()?;
+
+        let remote_head = remote_client.get_head()?;
+        d(&format!("remote head {:?}", remote_head));
+
+        let locally_known_remote_head = SyncMetadataClient::remote_head(&db_tx)?;
+        d(&format!("local head {:?}", locally_known_remote_head));
+
+        // Check if the server is empty - populate it.
+        if remote_head == Uuid::nil() {
```

Comment:
I'm expecting to see, at some point, a little state enum pop up here. After all, we're encapsulating a bunch of potential states:

- Both sides are at the same head.
- The remote server is empty, and we are at head H.
- The remote server is not empty, and we are empty.
- The remote server is at head HO and we are at HN, where HN descends from HO.
- … and vice versa.
- The remote server is at head HR, we are at head HL, and we have a shared parent HS.

Breaking this into an enum allows us to separate the "where are we?" diagnostic stage (fetch the local head, fetch the remote head, figure out parents…) from the "now what do we do?" step, which can then be neatly laid out in a `match` on the enum.

Follow-up is fine.